cppcache/src/TcrConnection.cpp

```diff
@@ -899,154 +915,173 @@ char* TcrConnection::readMessage(size_t* recvLen,
+      receiveBuffer, CHUNK_HEADER_LENGTH);
+  header.chunkLength = input.readInt32();
+  header.flags = input.read();
+  LOGDEBUG(
+      "TcrConnection::readChunkHeader: "
+      ", chunkLen=%" PRId32 ", lastChunkAndSecurityFlags=0x%" PRIx8,
+      header.chunkLength, header.flags);
 
-      LOGDEBUG(
-          "TcrConnection::readMessageChunked: received chunk body %d "
-          "from endpoint %s; bytes: %s",
-          chunkNum, m_endpoint,
-          Utils::convertBytesToString(chunk_body, chunkLen).c_str());
-      // Process the chunk; the actual processing is done by a separate thread
-      // ThinClientBaseDM::m_chunkProcessor.
-
-      reply.processChunk(chunk_body, chunkLen,
-                         m_endpointObj->getDistributedMemberID(), isLastChunk);
-    } while (!(isLastChunk & 0x1));
-  } catch (const Exception&) {
-    auto ex = reply.getChunkedResultHandler()->getException();
-    LOGDEBUG("Found existing exception ", ex->what());
-    reply.getChunkedResultHandler()->clearException();
-    throw;
+  return header;
+}
+
+std::vector<uint8_t> TcrConnection::readChunkBody(
+    std::chrono::microseconds timeout, int32_t chunkLength) {
+  std::vector<uint8_t> chunkBody(chunkLength);
+  auto error = receiveData(reinterpret_cast<char*>(chunkBody.data()),
+                           chunkLength, timeout, true, false);
+  if (error != CONN_NOERR) {
+    if (error & CONN_TIMEOUT) {
+      throwException(
+          TimeoutException("TcrConnection::readChunkBody: "
+                           "connection timed out while receiving chunk body"));
+    } else {
+      throwException(
+          GeodeIOException("TcrConnection::readChunkBody: "
+                           "connection failure while receiving chunk body"));
+    }
   }
 
-  LOGFINER(
-      "TcrConnection::readMessageChunked: read full reply "
-      "from endpoint %s",
-      m_endpoint);
+  LOGDEBUG(
+      "TcrConnection::readChunkBody: received chunk body from endpoint "
+      "%s; bytes: %s",
+      m_endpoint,
+      Utils::convertBytesToString(chunkBody.data(), chunkLength).c_str());
+  return chunkBody;
+}
+
+bool TcrConnection::processChunk(TcrMessageReply& reply,
+                                 std::chrono::microseconds timeout,
+                                 int32_t chunkLength,
+                                 int8_t lastChunkAndSecurityFlags) {
+  // NOTE: this buffer is allocated by readChunkBody, and reply.processChunk
+  // takes ownership, so we don't delete it here on failure
+  std::vector<uint8_t> chunkBody = readChunkBody(timeout, chunkLength);
+
+  // Process the chunk; the actual processing is done by a separate thread
+  // ThinClientBaseDM::m_chunkProcessor.
+  reply.processChunk(chunkBody, chunkLength,
+                     m_endpointObj->getDistributedMemberID(),
+                     lastChunkAndSecurityFlags);
+  return (lastChunkAndSecurityFlags & LAST_CHUNK_MASK) ? false : true;
```

Comment:
Any way to make this more self-documenting for a future developer?

Reply:
Added a comment, you're right this one's a little funny-looking